Potential buffer leak in wrap io.netty.handler.ssl.SslHandler.wrap

We are seeing some leaks 
```
LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records: 
Created at:
	io.netty.buffer.AdaptiveByteBufAllocator.newDirectBuffer(AdaptiveByteBufAllocator.java:67)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:168)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:159)
	io.netty.buffer.AbstractByteBufAllocator.buffer(AbstractByteBufAllocator.java:96)
	io.netty.handler.codec.http.HttpObjectEncoder.encodeInitHttpMessage(HttpObjectEncoder.java:523)
	io.netty.handler.codec.http.HttpObjectEncoder.encodeJustHttpMessage(HttpObjectEncoder.java:199)
	io.netty.handler.codec.http.HttpObjectEncoder.encode(HttpObjectEncoder.java:186)
	io.netty.handler.codec.http.HttpClientCodec$Encoder.encode(HttpClientCodec.java:290)
	io.netty.handler.codec.http.HttpObjectEncoder.write(HttpObjectEncoder.java:105)
	io.netty.channel.CombinedChannelDuplexHandler.write(CombinedChannelDuplexHandler.java:344)
	io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:798)
	io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:737)
	io.netty.handler.timeout.IdleStateHandler.write(IdleStateHandler.java:304)
	io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:798)
	io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:737)
	io.netty.handler.logging.LoggingHandler.write(LoggingHandler.java:288)
	io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:798)
	io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:776)
	io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:859)
	software.amazon.awssdk.http.nio.netty.internal.nrs.HttpStreamsHandler.unbufferedWrite(HttpStreamsHandler.java:327)
	software.amazon.awssdk.http.nio.netty.internal.nrs.HttpStreamsHandler.flushNext(HttpStreamsHandler.java:376)
	software.amazon.awssdk.http.nio.netty.internal.nrs.HttpStreamsHandler.write(HttpStreamsHandler.java:270)
	software.amazon.awssdk.http.nio.netty.internal.nrs.HttpStreamsClientHandler.write(HttpStreamsClientHandler.java:59)
	io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:798)
	io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:776)
	io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:859)
	io.netty.channel.DefaultChannelPipeline.writeAndFlush(DefaultChannelPipeline.java:1046)
	io.netty.channel.Channel.writeAndFlush(Channel.java:257)
	software.amazon.awssdk.http.nio.netty.internal.NettyRequestExecutor.writeRequest(NettyRequestExecutor.java:245)
	software.amazon.awssdk.http.nio.netty.internal.NettyRequestExecutor.makeRequest(NettyRequestExecutor.java:237)
	software.amazon.awssdk.http.nio.netty.internal.NettyRequestExecutor.lambda$makeRequestListener$10(NettyRequestExecutor.java:181)
	io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	io.netty.util.concurrent.PromiseTask.run(PromiseTask.java:106)
	io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:148)
	io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:141)
	io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:535)
	io.netty.channel.SingleThreadIoEventLoop.run(SingleThreadIoEventLoop.java:201)
	io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:1195)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	java.base/java.lang.Thread.run(Thread.java:1474)
```

It seems related to a leak in io.netty.handler.ssl.SslHandler.wrap evidenced by the following errors
```
Caused by: io.netty.util.internal.OutOfDirectMemoryError: failed to allocate 540672 byte(s) of direct memory (used: 12884787207, max: 12884901888)
	at io.netty.util.internal.PlatformDependent.incrementMemoryCounter(PlatformDependent.java:1102)
	at io.netty.util.internal.CleanerJava9$CleanableDirectBufferImpl.<init>(CleanerJava9.java:136)
	at io.netty.util.internal.CleanerJava9$CleanableDirectBufferImpl.<init>(CleanerJava9.java:131)
	at io.netty.util.internal.CleanerJava9.allocate(CleanerJava9.java:86)
	at io.netty.util.internal.PlatformDependent.allocateDirect(PlatformDependent.java:633)
	at io.netty.buffer.UnpooledDirectByteBuf.allocateDirectBuffer(UnpooledDirectByteBuf.java:129)
	at io.netty.buffer.UnpooledDirectByteBuf.<init>(UnpooledDirectByteBuf.java:70)
	at io.netty.buffer.UnpooledUnsafeDirectByteBuf.<init>(UnpooledUnsafeDirectByteBuf.java:50)
	at io.netty.buffer.UnsafeByteBufUtil.newDirectByteBuf(UnsafeByteBufUtil.java:734)
	at io.netty.buffer.AdaptiveByteBufAllocator$DirectChunkAllocator.allocate(AdaptiveByteBufAllocator.java:114)
	at io.netty.buffer.AdaptivePoolingAllocator$SizeClassChunkController.newChunkAllocation(AdaptivePoolingAllocator.java:734)
	at io.netty.buffer.AdaptivePoolingAllocator$Magazine.allocate(AdaptivePoolingAllocator.java:957)
	at io.netty.buffer.AdaptivePoolingAllocator$Magazine.tryAllocate(AdaptivePoolingAllocator.java:854)
	at io.netty.buffer.AdaptivePoolingAllocator$MagazineGroup.allocate(AdaptivePoolingAllocator.java:421)
	at io.netty.buffer.AdaptivePoolingAllocator.allocate(AdaptivePoolingAllocator.java:269)
	at io.netty.buffer.AdaptivePoolingAllocator.allocate(AdaptivePoolingAllocator.java:255)
	at io.netty.buffer.AdaptiveByteBufAllocator.newDirectBuffer(AdaptiveByteBufAllocator.java:67)
	at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:168)
	at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:159)
	at io.netty.handler.ssl.SslHandler$SslEngineType$1.allocateWrapBuffer(SslHandler.java:230)
	at io.netty.handler.ssl.SslHandler.allocateOutNetBuf(SslHandler.java:2390)
	at io.netty.handler.ssl.SslHandler.wrap(SslHandler.java:884)
	at io.netty.handler.ssl.SslHandler.wrapAndFlush(SslHandler.java:838)
	at io.netty.handler.ssl.SslHandler.flush(SslHandler.java:819)
	at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:760)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.flush(CombinedChannelDuplexHandler.java:529)
	at io.netty.channel.ChannelOutboundHandlerAdapter.flush(ChannelOutboundHandlerAdapter.java:125)
	at io.netty.channel.CombinedChannelDuplexHandler.flush(CombinedChannelDuplexHandler.java:354)
	at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:756)
	at io.netty.handler.logging.LoggingHandler.flush(LoggingHandler.java:304)
	at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:756)
	at io.netty.channel.ChannelDuplexHandler.flush(ChannelDuplexHandler.java:127)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:817)
	at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:776)
	at software.amazon.awssdk.http.nio.netty.internal.utils.DelegatingChannelHandlerContext.writeAndFlush(DelegatingChannelHandlerContext.java:196)
	at software.amazon.awssdk.http.nio.netty.internal.utils.OrderedWriteChannelHandlerContext.lambda$writeAndFlush$2(OrderedWriteChannelHandlerContext.java:54)
	at software.amazon.awssdk.http.nio.netty.internal.utils.OrderedWriteChannelHandlerContext.lambda$doInOrder$4(OrderedWriteChannelHandlerContext.java:73)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:148)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:141)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:535)
	at io.netty.channel.SingleThreadIoEventLoop.run(SingleThreadIoEventLoop.java:201)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:1195)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	... 1 more
```

The code in [wrap](https://github.com/netty/netty/blob/da22048fd5ea72f6f98a36251544796bd85be60a/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L848)

takes out a buffer
```
                ByteBuf buf = wrapDataSize > 0 ?
                        pendingUnencryptedWrites.remove(alloc, wrapDataSize, promise) :
                        pendingUnencryptedWrites.removeFirst(promise);
```

after which it proceeds to process it. Part of the processing may be allocating new buffers, e.g., [here](https://github.com/netty/netty/blob/da22048fd5ea72f6f98a36251544796bd85be60a/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L879C29-L879C108)

```
if (out == null) {
   out = allocateOutNetBuf(ctx, readableBytes, buf.nioBufferCount() + numPackets);
}
```

if this fails because there is not enough direct memory the `buf` buffer is leaked.

There is handling below of SSLException
```
catch (SSLException e) {
   // Either wrapMultiple(...) or wrap(...) did throw. In this case we need to release the buffer
   // that we removed from pendingUnencryptedWrites before failing the promise and rethrowing it.
   // Failing to do so would result in a buffer leak.
   // See https://github.com/netty/netty/issues/14644
   //
   // We don't need to release out here as this is done in a finally block already.
   buf.release();
   promise.setFailure(e);
   throw e;
}
```

but not out of direct memory exceptions. This seems like an oversight and causes one temporary failure to cause further failures to be more likely due to a leak.

Fixed by https://github.com/netty/netty/pull/17059


> On Jul 6, 2026, at 11:59 AM, Aleš Bizjak ***@***.***> wrote:
> 
> 
> abizjak
>  created an issue 
> (netty/netty#17057)
>  <https://github.com/netty/netty/issues/17057>
> We are seeing some leaks
> 
> LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
> Recent access records: 
> Created at:
> 	io.netty.buffer.AdaptiveByteBufAllocator.newDirectBuffer(AdaptiveByteBufAllocator.java:67)
> 	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:168)
> 	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:159)
> 	io.netty.buffer.AbstractByteBufAllocator.buffer(AbstractByteBufAllocator.java:96)
> 	io.netty.handler.codec.http.HttpObjectEncoder.encodeInitHttpMessage(HttpObjectEncoder.java:523)
> 	io.netty.handler.codec.http.HttpObjectEncoder.encodeJustHttpMessage(HttpObjectEncoder.java:199)
> 	io.netty.handler.codec.http.HttpObjectEncoder.encode(HttpObjectEncoder.java:186)
> 	io.netty.handler.codec.http.HttpClientCodec$Encoder.encode(HttpClientCodec.java:290)
> 	io.netty.handler.codec.http.HttpObjectEncoder.write(HttpObjectEncoder.java:105)
> 	io.netty.channel.CombinedChannelDuplexHandler.write(CombinedChannelDuplexHandler.java:344)
> 	io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:798)
> 	io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:737)
> 	io.netty.handler.timeout.IdleStateHandler.write(IdleStateHandler.java:304)
> 	io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:798)
> 	io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:737)
> 	io.netty.handler.logging.LoggingHandler.write(LoggingHandler.java:288)
> 	io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:798)
> 	io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:776)
> 	io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:859)
> 	software.amazon.awssdk.http.nio.netty.internal.nrs.HttpStreamsHandler.unbufferedWrite(HttpStreamsHandler.java:327)
> 	software.amazon.awssdk.http.nio.netty.internal.nrs.HttpStreamsHandler.flushNext(HttpStreamsHandler.java:376)
> 	software.amazon.awssdk.http.nio.netty.internal.nrs.HttpStreamsHandler.write(HttpStreamsHandler.java:270)
> 	software.amazon.awssdk.http.nio.netty.internal.nrs.HttpStreamsClientHandler.write(HttpStreamsClientHandler.java:59)
> 	io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:798)
> 	io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:776)
> 	io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:859)
> 	io.netty.channel.DefaultChannelPipeline.writeAndFlush(DefaultChannelPipeline.java:1046)
> 	io.netty.channel.Channel.writeAndFlush(Channel.java:257)
> 	software.amazon.awssdk.http.nio.netty.internal.NettyRequestExecutor.writeRequest(NettyRequestExecutor.java:245)
> 	software.amazon.awssdk.http.nio.netty.internal.NettyRequestExecutor.makeRequest(NettyRequestExecutor.java:237)
> 	software.amazon.awssdk.http.nio.netty.internal.NettyRequestExecutor.lambda$makeRequestListener$10(NettyRequestExecutor.java:181)
> 	io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
> 	io.netty.util.concurrent.PromiseTask.run(PromiseTask.java:106)
> 	io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:148)
> 	io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:141)
> 	io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:535)
> 	io.netty.channel.SingleThreadIoEventLoop.run(SingleThreadIoEventLoop.java:201)
> 	io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:1195)
> 	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
> 	java.base/java.lang.Thread.run(Thread.java:1474)
> It seems related to a leak in io.netty.handler.ssl.SslHandler.wrap evidenced by the following errors
> 
> Caused by: io.netty.util.internal.OutOfDirectMemoryError: failed to allocate 540672 byte(s) of direct memory (used: 12884787207, max: 12884901888)
> 	at io.netty.util.internal.PlatformDependent.incrementMemoryCounter(PlatformDependent.java:1102)
> 	at io.netty.util.internal.CleanerJava9$CleanableDirectBufferImpl.<init>(CleanerJava9.java:136)
> 	at io.netty.util.internal.CleanerJava9$CleanableDirectBufferImpl.<init>(CleanerJava9.java:131)
> 	at io.netty.util.internal.CleanerJava9.allocate(CleanerJava9.java:86)
> 	at io.netty.util.internal.PlatformDependent.allocateDirect(PlatformDependent.java:633)
> 	at io.netty.buffer.UnpooledDirectByteBuf.allocateDirectBuffer(UnpooledDirectByteBuf.java:129)
> 	at io.netty.buffer.UnpooledDirectByteBuf.<init>(UnpooledDirectByteBuf.java:70)
> 	at io.netty.buffer.UnpooledUnsafeDirectByteBuf.<init>(UnpooledUnsafeDirectByteBuf.java:50)
> 	at io.netty.buffer.UnsafeByteBufUtil.newDirectByteBuf(UnsafeByteBufUtil.java:734)
> 	at io.netty.buffer.AdaptiveByteBufAllocator$DirectChunkAllocator.allocate(AdaptiveByteBufAllocator.java:114)
> 	at io.netty.buffer.AdaptivePoolingAllocator$SizeClassChunkController.newChunkAllocation(AdaptivePoolingAllocator.java:734)
> 	at io.netty.buffer.AdaptivePoolingAllocator$Magazine.allocate(AdaptivePoolingAllocator.java:957)
> 	at io.netty.buffer.AdaptivePoolingAllocator$Magazine.tryAllocate(AdaptivePoolingAllocator.java:854)
> 	at io.netty.buffer.AdaptivePoolingAllocator$MagazineGroup.allocate(AdaptivePoolingAllocator.java:421)
> 	at io.netty.buffer.AdaptivePoolingAllocator.allocate(AdaptivePoolingAllocator.java:269)
> 	at io.netty.buffer.AdaptivePoolingAllocator.allocate(AdaptivePoolingAllocator.java:255)
> 	at io.netty.buffer.AdaptiveByteBufAllocator.newDirectBuffer(AdaptiveByteBufAllocator.java:67)
> 	at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:168)
> 	at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:159)
> 	at io.netty.handler.ssl.SslHandler$SslEngineType$1.allocateWrapBuffer(SslHandler.java:230)
> 	at io.netty.handler.ssl.SslHandler.allocateOutNetBuf(SslHandler.java:2390)
> 	at io.netty.handler.ssl.SslHandler.wrap(SslHandler.java:884)
> 	at io.netty.handler.ssl.SslHandler.wrapAndFlush(SslHandler.java:838)
> 	at io.netty.handler.ssl.SslHandler.flush(SslHandler.java:819)
> 	at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:760)
> 	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.flush(CombinedChannelDuplexHandler.java:529)
> 	at io.netty.channel.ChannelOutboundHandlerAdapter.flush(ChannelOutboundHandlerAdapter.java:125)
> 	at io.netty.channel.CombinedChannelDuplexHandler.flush(CombinedChannelDuplexHandler.java:354)
> 	at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:756)
> 	at io.netty.handler.logging.LoggingHandler.flush(LoggingHandler.java:304)
> 	at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:756)
> 	at io.netty.channel.ChannelDuplexHandler.flush(ChannelDuplexHandler.java:127)
> 	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:817)
> 	at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:776)
> 	at software.amazon.awssdk.http.nio.netty.internal.utils.DelegatingChannelHandlerContext.writeAndFlush(DelegatingChannelHandlerContext.java:196)
> 	at software.amazon.awssdk.http.nio.netty.internal.utils.OrderedWriteChannelHandlerContext.lambda$writeAndFlush$2(OrderedWriteChannelHandlerContext.java:54)
> 	at software.amazon.awssdk.http.nio.netty.internal.utils.OrderedWriteChannelHandlerContext.lambda$doInOrder$4(OrderedWriteChannelHandlerContext.java:73)
> 	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:148)
> 	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:141)
> 	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:535)
> 	at io.netty.channel.SingleThreadIoEventLoop.run(SingleThreadIoEventLoop.java:201)
> 	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:1195)
> 	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
> 	... 1 more
> The code in wrap <https://github.com/netty/netty/blob/da22048fd5ea72f6f98a36251544796bd85be60a/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L848>
> takes out a buffer
> 
>                 ByteBuf buf = wrapDataSize > 0 ?
>                         pendingUnencryptedWrites.remove(alloc, wrapDataSize, promise) :
>                         pendingUnencryptedWrites.removeFirst(promise);
> after which it proceeds to process it. Part of the processing may be allocating new buffers, e.g., here <https://github.com/netty/netty/blob/da22048fd5ea72f6f98a36251544796bd85be60a/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L879C29-L879C108>
> if (out == null) {
>    out = allocateOutNetBuf(ctx, readableBytes, buf.nioBufferCount() + numPackets);
> }
> if this fails because there is not enough direct memory the buf buffer is leaked.
> 
> There is handling below of SSLException
> 
> catch (SSLException e) {
>    // Either wrapMultiple(...) or wrap(...) did throw. In this case we need to release the buffer
>    // that we removed from pendingUnencryptedWrites before failing the promise and rethrowing it.
>    // Failing to do so would result in a buffer leak.
>    // See https://github.com/netty/netty/issues/14644
>    //
>    // We don't need to release out here as this is done in a finally block already.
>    buf.release();
>    promise.setFailure(e);
>    throw e;
> }
> but not out of direct memory exceptions. This seems like an oversight and causes one temporary failure to cause further failures to be more likely due to a leak.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/netty/netty/issues/17057?email_source=notifications&email_token=AADLIQXTN6EITYBBGGBI2LT5DN2ALA5CNFSL4Z3JMQ5C6L3HNF2C22DVMIXUS43TOVSS6NBYGE4DAMJQHA2TDJTSMVQXG33OVJZXKYTTMNZGSYTFMSSWK5TFNZ2KYZTPN52GK4S7MNWGSY3L>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AADLIQXSYBSUHKGYH3RGWC35DN2ALAVCNFSNUABDKJSXA33TNF2G64TZHMYTANRUGU3DGO2JONZXKZJ3GQ4DCOBQGEYDQNJRUF3AE>.
> You are receiving this because you are subscribed to this thread.
>

--- a/handler/src/main/java/io/netty/handler/ssl/SslHandler.java
+++ b/handler/src/main/java/io/netty/handler/ssl/SslHandler.java
@@ -885,10 +885,10 @@ public class SslHandler extends ByteToMessageDecoder implements ChannelOutboundH
                         }
                         result = wrap(alloc, engine, buf, out);
                     }
-                } catch (SSLException e) {
-                    // Either wrapMultiple(...) or wrap(...) did throw. In this case we need to release the buffer
-                    // that we removed from pendingUnencryptedWrites before failing the promise and rethrowing it.
-                    // Failing to do so would result in a buffer leak.
+                } catch (Throwable e) {
+                    // Either wrapMultiple(...), wrap(...) or allocateOutNetBuf(...) did throw.
+                    // In this case we need to release the buffer that we removed from pendingUnencryptedWrites
+                    // before failing the promise and rethrowing it. Failing to do so would result in a buffer leak.
                     // See https://github.com/netty/netty/issues/14644
                     //
                     // We don't need to release out here as this is done in a finally block already.

